src/array/cuda/rowwise_sampling_prob.cu

```diff
@@ -0,0 +1,639 @@
+    // with replacement.
+    _CSRRowWiseSampleReplaceKernel<IdType, FloatType, BLOCK_CTAS, TILE_SIZE><<<grid, block, 0, stream>>>(
+      rand_seed, 
+      num_picks, 
+      num_rows, 
+      slice_rows, 
+      in_ptr, 
+      in_cols, 
+      data, 
+      prob_data, 
+      out_ptr, 
+      out_rows, 
+      out_cols, 
+      out_idxs, 
+      temp_ptr, 
+      temp);
+    device->FreeWorkspace(ctx, temp);
+  } else {
+    IdType* temp_idxs = static_cast<IdType*>(
+      device->AllocWorkspace(ctx, (temp_len) * sizeof(IdType)));
+
+    constexpr int BLOCK_CTAS = 128 / CTA_SIZE;
+    // the number of rows each thread block will cover
+    constexpr int TILE_SIZE = BLOCK_CTAS;
+    const dim3 block(CTA_SIZE, BLOCK_CTAS);
+    const dim3 grid((num_rows + TILE_SIZE - 1) / TILE_SIZE);
+
+    // Compute A-Res value. A-Res value needs to be calculated only if deg 
+    // is greater than num_picks in without replacement weighted rowwise sampling.
+    _CSR_A_Res<IdType, FloatType, BLOCK_CTAS, TILE_SIZE><<<grid, block, 0, stream>>>(
+      rand_seed, 
+      num_picks, 
+      num_rows, 
+      slice_rows, 
+      in_ptr, 
+      prob_data, 
+      temp_ptr, 
+      temp_idxs, 
+      temp);
+
+    // sort A-Res value array.
+    FloatType* sort_temp = static_cast<FloatType*>(
+      device->AllocWorkspace(ctx, temp_len * sizeof(FloatType)));
+    IdType* sort_temp_idxs = static_cast<IdType*>(
+      device->AllocWorkspace(ctx, (temp_len) * sizeof(IdType)));
+
+    cub::DoubleBuffer<FloatType> sort_keys(temp, sort_temp);
+    cub::DoubleBuffer<IdType> sort_values(temp_idxs, sort_temp_idxs);
+
+    void *d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+
+    CUDA_CALL(cub::DeviceSegmentedRadixSort::SortPairsDescending(
+      d_temp_storage, 
+      temp_storage_bytes, 
+      sort_keys, 
+      sort_values, 
+      temp_len, 
+      num_rows, 
+      temp_ptr, 
+      temp_ptr + 1, 
+      0, 
+      sizeof(FloatType) * 8, stream));
+
+    d_temp_storage = device->AllocWorkspace(ctx, temp_storage_bytes);
+
+    CUDA_CALL(cub::DeviceSegmentedRadixSort::SortPairsDescending(
+      d_temp_storage, 
+      temp_storage_bytes, 
+      sort_keys, 
+      sort_values, 
+      temp_len, 
+      num_rows, 
+      temp_ptr, 
+      temp_ptr + 1, 0, 
+      sizeof(FloatType) * 8, 
+      stream));
+
+    device->FreeWorkspace(ctx, d_temp_storage);
+    device->FreeWorkspace(ctx, temp);
+    device->FreeWorkspace(ctx, temp_idxs);
+    device->FreeWorkspace(ctx, sort_temp);
+    // Perform weighted row-wise sampling on a CSR matrix, and generate a COO matrix,
+    // without replacement.
+    _CSRRowWiseSampleKernel<IdType, FloatType, BLOCK_CTAS, TILE_SIZE><<<grid, block, 0, stream>>>(
+      num_picks,
+      num_rows,
+      slice_rows,
+      in_ptr,
+      in_cols,
+      data, 
+      out_ptr,
+      out_rows,
+      out_cols,
+      out_idxs, 
+      temp_ptr,
+      sort_temp_idxs
```

Comment:
It may be appropriate to use sort_values.Current() instead of sort_temp_idxs,and call   device->FreeWorkspace(ctx, sort_temp) after the kernel.

Reply:
You're right.

Reply:
Thank you for pointing out this error, which has been fixed in #4064 pr